### PR TITLE
feat(Makefile): add `make format`

### PR DESCRIPTION
**Pull request by Benjamin-eecs**
## Description

#220 

- add flake8 configuration
- modify yapf configuration to resolve conflict with flake8
- `make format` using yapf, isort and ruff

## Roadmap

- [x] Support Linux and MacOS
- [x] `make format`

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	$(PYTHON) -m isort --project $(PROJECT_PATH) --check $(PYTHON_FILES) && \
+	$(PYTHON) -m black --check $(PYTHON_FILES) tutorials
+
+ruff: ruff-install
+	$(PYTHON) -m ruff check .
+
+ruff-fix: ruff-install
+	$(PYTHON) -m ruff check . --fix --exit-non-zero-on-fix
+
+mypy: mypy-install
```
Comment (Obs01ete) [inline Makefile:121]: We also need `--namespace-packages` key

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+pytest-install:
+	$(call check_pip_install,pytest)
+	$(call check_pip_install,pytest-cov)
+	$(call check_pip_install,pytest-xdist)
+
+test-install: pytest-install
+	$(PYTHON) -m pip install --requirement tests/requirements.txt
+
+go-install:
```
Comment (Obs01ete) [inline Makefile:86]: Why do we need go?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	# requires go >= 1.16
+	command -v go || (sudo apt-get install -y golang && sudo ln -sf /usr/lib/go/bin/go /usr/bin/go)
+
+addlicense-install: go-install
+	command -v addlicense || go install github.com/google/addlicense@latest
+
+# Tests
+
+pytest: test-install
```
Comment (Obs01ete) [inline Makefile:95]: This way to get coerage works better:
```bash
coverage erase
coverage run --source=. -m pytest .
coverage html
# Open htmlcov/index.html
```
It covers code that is not tested at all.

File: camel/agents/__init__.py [hunk 4]
```diff
@@ -13,11 +13,11 @@
 # =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
 from .base import BaseAgent
 from .chat_agent import ChatAgent, ChatAgentResponse
-from .task_agent import TaskPlannerAgent, TaskSpecifyAgent
```
Comment (Obs01ete) [inline camel/agents/__init__.py:16]: Why are imports reshuffled? Do you use sort differently from what is there in the pre commit hook?

**Review by Obs01ete — COMMENTED** [review]
Overall great to have all tools in a Makefile. The only question is why py files are affected

File: camel/agents/__init__.py (hunk 4 above)
Comment (Benjamin-eecs) [reply camel/agents/__init__.py:16]: actually, using `make format` actually helps me find several circular import bug, will fix in this pr too

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+pylint: pylint-install
+	$(PYTHON) -m pylint $(PROJECT_PATH)
+
+flake8: flake8-install
+	$(PYTHON) -m flake8 --count --show-source --statistics
+
+py-format: py-format-install
+	$(PYTHON) -m isort --project $(PROJECT_PATH) --check $(PYTHON_FILES) && \
+	$(PYTHON) -m black --check $(PYTHON_FILES) tutorials
```
Comment (Obs01ete) [inline Makefile:96]: Please revert to yapf. I do not see how black is better to the extent that we migrate to it.

**Review by Obs01ete — CHANGES_REQUESTED** [review]

File: Makefile (hunk 5 above)
Comment (Benjamin-eecs) [reply Makefile:96]: Also, inconsistency using `''` and `""`

Comment (Obs01ete) [reply Makefile:96]: The only question is if you want to have you Makefile merged early on and then continue the infinite discussion about style preferences? @Benjamin-eecs

Comment (Benjamin-eecs) [reply Makefile:96]: sure, get makefile intergration merged then

**Comment by Benjamin-eecs** [comment]
I have decided to close this PR and will split the changes into several smaller, separate PRs. This will make it easier for the reviewers to manage and ensure thorough review of each change.

**Pull request closed without merging**